waspc/src/Wasp/AppSpec/ExtImport.hs

```diff
@@ -25,11 +24,21 @@ data ExtImport = ExtImport
 
 instance FromJSON ExtImport where
   parseJSON = withObject "ExtImport" $ \o -> do
-    name' <- o .: "name"
+    kindStr <- o .: "kind"
+    nameStr <- o .: "name"
     pathStr <- o .: "path"
-    case parseRelFileP pathStr of
-      Just path' -> pure $ ExtImport {name = name', path = path'}
-      Nothing -> fail $ "Failed to parse relative posix path to file: " <> pathStr
+    extImportName <- parseExtImportName kindStr nameStr
+    extImportPath <- parseExtImportPath pathStr
+    return $ ExtImport extImportName extImportPath
+    where
+      parseExtImportName kindStr nameStr = case kindStr of
+        "default" -> pure $ ExtImportModule nameStr
+        "named" -> pure $ ExtImportField nameStr
+        _ -> fail $ "Failed to parse import kind: " <> kindStr
```

Comment:
Hm I am not sure I get what you would like to have here.

What I would imagine I might want here is to make sure I have covered all the possible values of `ExtImportName`. So if I add a new data constructor, I would love compiler to warn me that I don't handle parsing that produces this new data constructor.

You however don't seem to have that in the TS playground link you shared. What you have is defining "default" and "named" in one place and then compiler telling you if one of them was misspelled in the `switch`. You don't even have the exaustive pattern check (that all were covered) in your `switch`, but I believe you can easily implement that via Record or via that trick in the `default` with non-reachable code. Let's say you have that -> I still don't see what you got with all that. We define `default` and `named` strings in just a single place here, so I don't see the use in externalizing that definition.

As for how to possibly get what I mentioned above, I imagine we could write something like this:
```hs
data ExtImportNameKind = Default | Named deriving (..., FromJSON)

extImportNameKind :: ExtImportName -> ExtImportNameKind
extImportNameKind = \case
  ExtImportModule {} -> Default
  ExtImportField {} -> Named

-- Here is your `parseExtImportName` function, but slightly changed.
-- And up there in parser you would have `kind <- o .: "kind"`, instead
-- of `kindStr <- o .: "kind"`.
parseExtImportName kind nameStr = case kind of
  Default -> pure $ ExtImportModule nameStr
  Named -> pure $ ExtImportField nameStr
```

This would now do what I want -> if we add a new ExtImportName data constructor, we would get a compiler error. I think it also does what you want (although I was not sure what you want)? It does achieve this in a bit clumsily way though: now we have this function `extImportNameKind` that I never use anywhere except to get compiler errors on missing cases. So if you added new data constructor to `ExtImportName`, first you would get error in `extImportNameKind` that case is missing. Then would add that case, and while doing that, you would add new data constructor to `data ExtImportNameKind`, since you would need it. Then, you wuold get an error in `parseExtImportName`, and finally you would add a handler there for the case when kind is that new kind. So it does what we need!

I have to admit it took me too long to get here hehe, but it was fun.

Apologies if I missed somethinig on your reasoning for untagged instances, please explain it more to me.

Reply:
Yeah, you're right. I shared the wrong snippet. I can't remember what I wanted to share though.

Your experiment is pretty cool, but yeah, it's a little clumsy.

Anyway, what I wanted to show/request was much simpler, I don't know what I was thinking with that snippet.

```typescript
declare function iWorkWithImportKinds(imp: ImportKind): void
declare function iWorkWithStrings(s: string): void

type ImportKind = "default" | "named"

const kind: ImportKind = "default"

// no need for any wrapping/unwrapping.
iWorkWithImportKinds(kind)
iWorkWithStrings(kind)
```
I guess what I want is subtyping (it's not that inherent to union types after all). From what I got, `DataKinds` does something like this?

Reply:
Hm I don't quite get what you want I ahve to admit, seems like what you have in this new snippet doesn't have much to do with the intiail conversation to me?

But I also forgot quite a bit so maybe it doesn't matter.

`DataKinds` allows you to define your own kinds. So instead of just `*` an `* -> *` and similar, you can now define your own, and you can use it in some places where it is allowed to specify kinds, e.g. `data Foo (f :: BarKind) = Foo String` -> now this `f` shadow param can be only one of types that belong to the kind `BarKind`. This is just a simple example.